### PR TITLE
Add multi argument callback helpers.

### PR DESCRIPTION
**Pull request by crocket**
This pull request does several things.
- It makes sure that line widths don't exceed 80 in GHCJS/Foreign/Callback.hs
- It adds jsbits/callback.js which contains `h$makeCallbackMulti`.
- It adds `GHCJS/Foreign/Callback/Variadic.hs` which enables polyvariadic callback functions.
- It adds `syncCallbackMulti`, `syncCallbackMulti'`, and `asyncCallbackMulti` to
  GHCJS/Foreign/Callback.hs. These functions utilize `h$makeCallbackMulti` and `VariadicCallback` for convenience of users. Refer to README for details.
- It bumps up the version to 0.2.1.0
- It documents `OnBlocked` and `syncCallback'` more clearly on README.
- It documents `syncCallbackMulti`, `syncCallbackMulti'`, and
  `asyncCallbackMulti` on README.

I tested `syncCallbackMulti`, `syncCallbackMulti'`, and `asyncCallbackMulti`
against the NodeJS example I included in README, and the functions worked well in my example.

This closes https://github.com/ghcjs/ghcjs-base/issues/76

It doesn't break existing functions.
Thus, I want this pull request to be merged as soon as possible if maintainers don't have major objections.

### CODE REVIEW

**Comment by hamishmack** [comment]
LGTM

File: GHCJS/Foreign/Callback.hs [hunk 1]
```diff
@@ -117,28 +140,61 @@ syncCallback3' :: (JSVal -> JSVal -> JSVal -> IO JSVal)
...
+     referenced by the IO action.
+ -}
+asyncCallback :: IO () -- the action that the callback runs
+              -> IO (Callback (IO ())) -- the callback
 asyncCallback x = js_asyncCallback (unsafeCoerce x)
 
-asyncCallback1 :: (JSVal -> IO ())            -- ^ the function that the callback calls
-               -> IO (Callback (JSVal -> IO ())) -- ^ the calback
+asyncCallback1 :: (JSVal -> IO ()) -- the function that the callback calls
+               -> IO (Callback (JSVal -> IO ())) -- the calback
```
Comment (crocket) [inline GHCJS/Foreign/Callback.hs]: Sorry, I haven't learned haddock, yet. I'll restore haddock markup cues.

**Comment by crocket** [comment]
@hamishmack Would it be ok to replace `JavaScript.Array.toListIO :: SomeJSArray m -> IO [JSVal]` with `GHCJS.Prim.fromJSArray :: JSVal -> IO [JSVal]` in my multi-argument callback functions as below?

```haskell
{- | Make a callback (JavaScript function) that runs the supplied IO function
     in a synchronous thread when called. The callback takes an arbitrary
     number of arguments that it passes as an array of JSVal values to the
     Haskell function.
     Call 'releaseCallback' on the callback when done with the callback,
     freeing data referenced by the function.
 -}
syncCallbackMulti :: OnBlocked -- ^ what to do when the thread blocks
                  -> ([JSVal] -> IO ()) -- ^ the Haskell action
                  -> IO (Callback ([JSVal] -> IO ())) -- ^ the callback
syncCallbackMulti onBlocked f = do
  js_syncCallbackMulti (onBlocked == ContinueAsync) $ unsafeCoerce $ \args ->
    Array.toListIO (unsafeCoerce args) >>= f
```

**Comment by hamishmack** [comment]
I think it would be fine.

**Comment by crocket** [comment]
@hamishmack @luite I think the pull request is ready for merge. I request that a tag, `v0.2.1.0` be added to the git repository after merging this commit. Is it ok for you to do so?

**Pull request closed without merging**